radix-engine/src/kernel/call_frame.rs

```diff
@@ -24,13 +22,17 @@ use super::kernel_api::LockInfo;
 pub struct Message {
     pub copy_references: Vec<NodeId>,
     pub move_nodes: Vec<NodeId>,
+    pub copy_transient_references: Vec<NodeId>,
+    pub copy_direct_references: Vec<NodeId>,
```

Comment:
What's the relationship between `copy_references`, `copy_transient_references` and `copy_direct_references`?